### PR TITLE
Avoid roundtrip [fconstr] -> [constr] -> [fconstr] in the kernel

### PR DESCRIPTION
**Pull request by gmalecha**
This is based on the discussion here: https://coq.zulipchat.com/#narrow/stream/237656-Coq-devs-.26-plugin-devs/topic/reduction.20in.20the.20kernel

It should be completely backwards compatible. I would like to get a benchmark run to determine if there is any impact. The code is a hackish right now (and contains some extra comments) just to test the potential performance difference, I will clean up before merge if we determine if it is useful.

- [ ] Added / updated **test-suite**.

<!-- Pointers to relevant developer documentation:

Contributing guide: https://github.com/coq/coq/blob/master/CONTRIBUTING.md

Test-suite: https://github.com/coq/coq/blob/master/test-suite/README.md

Changelog: https://github.com/coq/coq/blob/master/doc/changelog/README.md

Building the doc: https://github.com/coq/coq/blob/master/doc/README.md
Sphinx: https://github.com/coq/coq/blob/master/doc/sphinx/README.rst
doc_gram: https://github.com/coq/coq/blob/master/doc/tools/docgram/README.md

Overlays: https://github.com/coq/coq/blob/master/dev/ci/user-overlays/README.md

### CODE REVIEW

**Comment by SkySkimmer** [comment]
@coqbot bench

File: kernel/reduction.ml [hunk 1]
```diff
@@ -915,16 +917,31 @@ let gen_conv cv_pb ?(l2r=false) ?(reds=TransparentState.full) env ?(evars=defaul
...
         ()
 
 let conv = gen_conv CONV
 let conv_leq = gen_conv CUMUL
 
+
+let gen_conv_fconstr cv_pb ?(l2r=false) ?(reds=TransparentState.full) env ?(evars=default_evar_handler) t1 t2 =
+  let univs = Environ.universes env in
+  let b =
+    if cv_pb = CUMUL then leq_constr_univs univs t1 (term_of_fconstr t2)
```
Comment (ppedrot) [inline kernel/reduction.ml:930]: Isn't this `term_of_fconstr` defeating the purpose?

Comment (gmalecha) [reply kernel/reduction.ml:930]: If I understand correctly, it is not *completely* defeating the purpose. My understanding is that this is doing a syntactic equality check (which I assume is potentially faster than the conversion check). The actual conversion check will still use the original `fconstr` which has (potentially) useful reduction information already inside of it.

I am experimenting with removing the conversion entirely.

Comment (gmalecha) [reply kernel/reduction.ml:930]: Here's a non-scientific performance comparison:

| version | foo | foo2 |
|------------|-----|-------|
| master | 0.266s | 0.001s |
| with `term_of_fconstr` | 0.214s | 0.001s |
| no `term_of_fconstr` | 0.189s | 0.001s |

Here's the test code.

```coq
Fixpoint get_unit (n : nat) : Type :=
  match n with
  | 0 => unit
  | S n => get_unit n
  end.

Fixpoint make (n : nat) : Type :=
  match n with
  | 0 => unit
  | S n => get_unit n -> make n
  end.

Fixpoint make2 (n : nat) : Type :=
  match n with
  | 0 => unit
  | S n => unit -> make2 n
  end.

Definition foo : make 1000.
  compute. intros. apply tt.
Qed.

Definition foo2 : make2 1000.
  compute. intros. apply tt.
Qed.

Lemma xxx : unit.
Proof.
  apply foo; exact tt.
Time Qed.

Lemma xxx2 : unit.
Proof.
  apply foo2; exact tt.
Time Qed.
```

**Comment by gmalecha** [comment]
@SkySkimmer Can you re-run the bench with the new changes to completely remove the syntactic equality check?

**Comment by SkySkimmer** [comment]
@coqbot bench

**Comment by SkySkimmer** [comment]
Not sure how useful it wil be with this many failures

**Comment by gmalecha** [comment]
Thanks. I'm going to try to look into the failures, they are all quite mysterious to me.

**Comment by gmalecha** [comment]
The latest change fixed most of the issues. I'm not sure about fiat-crypto, but unimath is a timeout. This seems to suggest that the cost of maintaining sharing is too high compared to the (hypothetical) benefit. It might be possible to regain some efficiency with a lazy copy.

**Comment by SkySkimmer** [comment]
@coqbot bench

**Comment by Alizter** [comment]
Those slowdowns in prennial are looking pretty worrying. I wonder why they are happening.
```
│   OLD      NEW     DIFF     %DIFF    Ln                 FILE                                                          │
│ 43.4810  67.2740  23.7930   54.72%   521  coq-perennial/src/program_proof/txn/twophase_refinement_proof.v.html        │
│  1.8600  16.4640  14.6040  785.16%   988  coq-perennial/src/program_proof/buf/buf_proof.v.html                        │
│ 28.0570  41.7860  13.7290   48.93%  2290  coq-perennial/src/goose_lang/logical_reln_fund.v.html     
```

**Comment by gmalecha** [comment]
@Alizter the issue is that the copy of the `fconstr` is too expensive to be justified by the savings from sharing in most cases. I don't think this will get merged unless there is a way to speed that up a lot.

**Comment by ppedrot** [comment]
This is indeed starting to look like a failed experiment...

**Comment by gmalecha** [comment]
@ppedrot I concur. My current take is that `copy_fconstr` is much more expensive than `inject (to_constr ..)`. The later loses sharing, the former does not. Do you think there could be a way to speed up the copy operation either by making it lazier or by memoizing larger-grained pieces of the copy?

**Comment by ppedrot** [comment]
Closing this, since this PR seems to fulfilled its goal.

**Pull request closed without merging**